reactive/etcd.py

```diff
@@ -786,6 +800,78 @@ def remove_nrpe_config(nagios=None):
         nrpe_setup.remove_check(shortname=service)
 
 
+@when('endpoint.prometheus.joined',
+      'leadership.is_leader',
+      'certificates.ca.available')
+def register_prometheus_jobs():
+    log('Registering Prometheus metrics collection.')
+    prometheus = endpoint_from_flag('endpoint.prometheus.joined')
+    cluster = endpoint_from_flag('cluster.joined')
+
+    if not etcd_reachable_from_endpoint(prometheus.endpoint_name):
+        log('Aborting Prometheus metrics collection. Etcd is not reachable by'
+            ' prometheus client', WARNING)
+        return
+
+    peer_ips = cluster.get_db_ingress_addresses() if cluster else []
+    peer_ips.append(get_ingress_address('db'))
+    targets = ["{}:{}".format(ip, config('port')) for ip in peer_ips]
+    log('Configuring Prometheus scrape targets: {}'.format(targets), DEBUG)
+    prometheus.register_job(job_name='etcd',
+                            job_data={
+                                'scheme': 'https',
+                                'static_configs': [
+                                    {'targets': targets},
+                                ]
+                            })
+    set_flag('prometheus.configured')
+
+
+@when('endpoint.prometheus.departed')
+def reset_prometheus_config():
+    clear_flag('prometheus.configured')
+    clear_flag('grafana.configured')
+
+
+@when(
+    "prometheus.configured",
+    "endpoint.grafana.joined",
+    "leadership.is_leader"
+)
+@when_not("grafana.configured")
+def register_grafana_dashboard():
+    log("Configuring grafana dashboard", level=hookenv.INFO)
+    grafana = endpoint_from_flag("endpoint.grafana.joined")
+    prometheus = endpoint_from_flag('endpoint.prometheus.joined')
+
+    if not prometheus:
```

Comment:
If this possible? `register_grafana_dashboard` handler is only triggered when `prometheus.configured` flag is set, which can only happen if `endpoint.prometheus.joined` is available. Shouldn't we leave `len(prometheus.relations)` fail because `NoneType has no "relations" attribute`? I may be missing more in-depth knowledge about Juju relations, if this could be an issue.

Reply:
This is perhaps bit of a paranoid approach. I wasn't sure if it's guaranteed that `@when(endpoint.prometheus.departed)` is executed first in case that prometheus relation is removed.